### PR TITLE
CapJitAddTypeSignature in WrapperCategory

### PR DESCRIPTION
**Pull request by mohamed-barakat**
for
* UnderlyingCell
* AsObjectInWrapperCategory
* AsMorphismInWrapperCategory

### CODE REVIEW

**Review by zickgraf — CHANGES_REQUESTED** [review]
Please use the new helper functions

* CapJitDataTypeOfCategory,
* CapJitDataTypeOfObjectCategory,
* CapJitDataTypeOfMorphismOfCategory.

For an example see OppositeCategory.

**Comment by mohamed-barakat** [comment]
> Please use the new helper functions
> 
> * CapJitDataTypeOfCategory,

I don't see where to use this one, probably you listed it for the sake of completeness.

> * CapJitDataTypeOfObjectCategory,
> * CapJitDataTypeOfMorphismOfCategory.

Done.

> For an example see OppositeCategory.

Thanks.

File: CAP/gap/WrapperCategory.gd [hunk 1]
```diff
@@ -97,6 +97,22 @@ DeclareAttribute( "UnderlyingCategory",
 DeclareAttribute( "UnderlyingCell",
         IsWrapperCapCategoryCell );
 
+##
+CapJitAddTypeSignature( "UnderlyingCell",
+        [ IsWrapperCapCategoryObject ],
+  function ( input_types )
+    
+    return CapJitDataTypeOfObjectOfCategory( ModelingCategory( input_types[1].category ) );
```
Comment (zickgraf) [inline CAP/gap/WrapperCategory.gd:105]:
```suggestion
    return CapJitDataTypeOfObjectOfCategory( ModelingCategory( input_types[1].category ) );
    
```
(missing empty line)

File: CAP/gap/WrapperCategory.gd [hunk 2]
```diff
@@ -97,6 +97,22 @@ DeclareAttribute( "UnderlyingCategory",
...
+end );
+
+##
+CapJitAddTypeSignature( "UnderlyingCell",
+        [ IsWrapperCapCategoryMorphism ],
+  function ( input_types )
+    
+    return CapJitDataTypeOfMorphismOfCategory( ModelingCategory( input_types[1].category ) );
```
Comment (zickgraf) [inline CAP/gap/WrapperCategory.gd:113]:
```suggestion
    return CapJitDataTypeOfMorphismOfCategory( ModelingCategory( input_types[1].category ) );
    
```
(missing empty line)

**Review by zickgraf — CHANGES_REQUESTED** [review]
> I don't see where to use this one, probably you listed it for the sake of completeness.

Exactly, I did not think about the details here :D

File: CAP/gap/WrapperCategory.gd (hunk 1 above)
Comment (mohamed-barakat) [reply CAP/gap/WrapperCategory.gd:105]: Done.

File: CAP/gap/WrapperCategory.gd (hunk 2 above)
Comment (mohamed-barakat) [reply CAP/gap/WrapperCategory.gd:114]: Done.

File: CAP/gap/WrapperCategory.gd [hunk 3]
```diff
@@ -110,6 +128,15 @@ DeclareAttribute( "UnderlyingCell",
 DeclareOperation( "AsObjectInWrapperCategory",
                   [ IsWrapperCapCategory, IsCapCategoryObject ] );
 
+##
+CapJitAddTypeSignature( "AsObjectInWrapperCategory",
+        [ IsCapCategory, IsCapCategoryObject ],
```
Comment (zickgraf) [inline CAP/gap/WrapperCategory.gd:133]:
```suggestion
        [ IsWrapperCapCategory, IsCapCategoryObject ],
```

File: CAP/gap/WrapperCategory.gd [hunk 4]
```diff
@@ -118,6 +145,15 @@ DeclareOperation( "AsObjectInWrapperCategory",
 DeclareOperation( "AsMorphismInWrapperCategory",
                   [ IsWrapperCapCategoryObject, IsCapCategoryMorphism, IsWrapperCapCategoryObject ] );
 
+##
+CapJitAddTypeSignature( "AsMorphismInWrapperCategory",
+        [ IsCapCategory, IsWrapperCapCategoryObject, IsCapCategoryMorphism, IsWrapperCapCategoryObject ],
```
Comment (zickgraf) [inline CAP/gap/WrapperCategory.gd:150]:
```suggestion
        [ IsWrapperCapCategory, IsWrapperCapCategoryObject, IsCapCategoryMorphism, IsWrapperCapCategoryObject ],
```

**Review by zickgraf — CHANGES_REQUESTED** [review]

File: CAP/gap/WrapperCategory.gd (hunk 3 above)
Comment (mohamed-barakat) [reply CAP/gap/WrapperCategory.gd]: Done.

File: CAP/gap/WrapperCategory.gd (hunk 4 above)
Comment (mohamed-barakat) [reply CAP/gap/WrapperCategory.gd]: Done.

**Review by zickgraf — APPROVED** [review]

**Pull request merged**